### PR TITLE
Setup timeout monitor for all CI builds

### PR DESCRIPTION
**Pull request by blindpirate**
Recently we have several timeouts in this build configuration: https://builds.gradle.org/viewType.html?buildTypeId=Gradle_Check_Platform_14_bucket11&tab=buildTypeHistoryList&state=failed

Checking the timeout builds, all test tasks have been finished successfully. So there's something else in the build hanging. Previously we set up a timeout monitor for each `IntegrationTest`, now we set the monitor for the whole build. 

Also fixes https://github.com/gradle/gradle-private/issues/3051

### CODE REVIEW

**Comment by blindpirate** [comment]
@bot-gradle test QFL

File: buildSrc/subprojects/plugins/src/main/kotlin/org/gradle/plugins/lifecycle/LifecyclePlugin.kt [hunk 1]
```diff
@@ -97,6 +104,46 @@ class LifecyclePlugin : Plugin<Project> {
...
+    private
+    fun Project.determineTimeoutMillis() =
+        if (isRequestedTask(compileAllBuild) || isRequestedTask(sanityCheck) || isRequestedTask(quickTest)) {
+            TimeUnit.MINUTES.toMillis(30)
+        } else {
+            TimeUnit.MINUTES.toMillis(165) // 2h45m
+        }
+
+    private
+    fun Project.findOutIntegrationTestClasspath(): FileCollection {
```
Comment (blindpirate) [inline buildSrc/subprojects/plugins/src/main/kotlin/org/gradle/plugins/lifecycle/LifecyclePlugin.kt:136]: I don't think of a good way to retrieve the classpath...

**Comment by blindpirate** [comment]
@bot-gradle test QF

File: buildSrc/subprojects/plugins/src/main/kotlin/org/gradle/plugins/lifecycle/LifecyclePlugin.kt (hunk 1 above)
Comment (jjohannes) [reply buildSrc/subprojects/plugins/src/main/kotlin/org/gradle/plugins/lifecycle/LifecyclePlugin.kt:136]: I think you only need the classpath for the `JavaProcessStackTracesMonitor` which is in the `:internalIntegTesting` project. So you need the runtimeClasspath of that project:

`project(":internalIntegTesting").configurations["runtimeClasspath"]`

I wonder if the JavaProcessStackTracesMonitor should go somewhere else. Is there a particular reason why it is in the internalIntegTest project? Could it maybe be code in buildSrc and part of the LifcyclePlugin? (I am just wondering. Reaching into a subproject of the build here does not look right. But we can improve that later.)

A better name for this method here is probably something like: `Project.javaProcessStackTracesMonitorClasspath()`

Comment (blindpirate) [reply buildSrc/subprojects/plugins/src/main/kotlin/org/gradle/plugins/lifecycle/LifecyclePlugin.kt:136]: The reason we keep it in `internalIntegTesting` is that it's also used in integration tests (when an integration test timeouts, we also want to get the stacktraces). If we move this to buildSrc, it needs to be passed to test classpath as well... so, dilemma.

**Comment by blindpirate** [comment]
@bot-gradle test QF

**Comment by blindpirate** [comment]
@bot-gradle test QF

**Comment by blindpirate** [comment]
Thanks @jjohannes !

File: buildSrc/subprojects/plugins/src/main/kotlin/org/gradle/plugins/lifecycle/LifecyclePlugin.kt (hunk 1 above)
Comment (blindpirate) [reply buildSrc/subprojects/plugins/src/main/kotlin/org/gradle/plugins/lifecycle/LifecyclePlugin.kt]: Got this error:

```
[04:23:26]W:	 [Step 1/2] Exception in thread "Timer-3" java.lang.IllegalStateException: The configuration :internalIntegTesting:runtimeClasspath was resolved from a thread not managed by Gradle.
[04:23:26]W:	 [Step 1/2] 	at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration.resolveToStateOrLater(DefaultConfiguration.java:565)
[04:23:26]W:	 [Step 1/2] 	at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration.access$1900(DefaultConfiguration.java:142)
[04:23:26]W:	 [Step 1/2] 	at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration$ConfigurationFileCollection.getSelectedArtifacts(DefaultConfiguration.java:1251)
[04:23:26]W:	 [Step 1/2] 	at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration$ConfigurationFileCollection.visitContents(DefaultConfiguration.java:1242)
[04:23:26]W:	 [Step 1/2] 	at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration$ConfigurationFileCollection.getFiles(DefaultConfiguration.java:1232)
[04:23:26]W:	 [Step 1/2] 	at org.gradle.api.internal.file.AbstractFileCollection.isEmpty(AbstractFileCollection.java:242)
[04:23:26]W:	 [Step 1/2] 	at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration.isEmpty(DefaultConfiguration.java:499)
[04:23:26]W:	 [Step 1/2] 	at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration_Decorated.isEmpty(Unknown Source)
[04:23:26]W:	 [Step 1/2] 	at org.gradle.api.internal.file.CompositeFileCollection.isEmpty(CompositeFileCollection.java:101)
[04:23:26]W:	 [Step 1/2] 	at org.gradle.process.internal.JavaExecHandleBuilder.addClassicJavaRunArgs(JavaExecHandleBuilder.java:116)
[04:23:26]W:	 [Step 1/2] 	at org.gradle.process.internal.JavaExecHandleBuilder.getAllJvmArgs(JavaExecHandleBuilder.java:101)
[04:23:26]W:	 [Step 1/2] 	at org.gradle.process.internal.JavaExecHandleBuilder.getAllArguments(JavaExecHandleBuilder.java:375)
[04:23:26]W:	 [Step 1/2] 	at org.gradle.process.internal.JavaExecHandleBuilder.getAllArguments(JavaExecHandleBuilder.java:371)
[04:23:26]W:	 [Step 1/2] 	at org.gradle.process.internal.JavaExecHandleBuilder.getEffectiveArguments(JavaExecHandleBuilder.java:385)
[04:23:26]W:	 [Step 1/2] 	at org.gradle.process.internal.AbstractExecHandleBuilder.build(AbstractExecHandleBuilder.java:148)
[04:23:26]W:	 [Step 1/2] 	at org.gradle.process.internal.DefaultJavaExecAction.execute(DefaultJavaExecAction.java:39)
[04:23:26]W:	 [Step 1/2] 	at org.gradle.process.internal.DefaultExecActionFactory.javaexec(DefaultExecActionFactory.java:161)
[04:23:26]W:	 [Step 1/2] 	at org.gradle.api.internal.project.DefaultProject.javaexec(DefaultProject.java:1145)
[04:23:26]W:	 [Step 1/2] 	at org.gradle.plugins.lifecycle.LifecyclePlugin$setupTimeoutMonitorOnCI$$inlined$apply$lambda$1.run(Timer.kt:149)
[04:23:26]W:	 [Step 1/2] 	at java.base/java.util.TimerThread.mainLoop(Timer.java:556)
[04:23:26]W:	 [Step 1/2] 	at java.base/java.util.TimerThread.run(Timer.java:506)
```

Comment (jjohannes) [reply buildSrc/subprojects/plugins/src/main/kotlin/org/gradle/plugins/lifecycle/LifecyclePlugin.kt]: Mhh. Could we do the "schedule" inside a task action? Where we resolve the classpath first and then use only the result inside the schduled action?
We would have to always excute the task early. Maybe similar to what we do for `killExistingProcessesStartedByGradle`?
This timeout monitor code _could_ potentially also go into the Cleanup plugin I suppose.

**Pull request merged**